Improves Axom's installation files

kennyweiss opened the pull request:
# Summary

- This PR adds an `axom-config` file to make it easier to install axom in CMake-based user applications
- It adds a configuration-based `host-config` file to the installed `using-with-cmake` example. This makes it easier to configure the example with the same compiler and TPLs that were used to build axom.
- It adds a `using-with-blt` example
- It adds `fmt` as a dependency to both of the above examples to check that our built-in TPLs are properly installed 

TODO:

- [x] Incorporate Axom's TPL dependencies directly into `axom-config`

kennyweiss commented on src/cmake/AxomConfig.cmake:
```diff
@@ -71,3 +71,49 @@ configure_file(
 
 install(FILES ${CMAKE_BINARY_DIR}/include/axom/config.hpp DESTINATION include/axom)
+
+#------------------------------------------------------------------------------
```
Much of this logic was adapted from `conduit`'s installation logic.  Thanks @cyrush !

kennyweiss commented on src/examples/using-with-cmake/host-config.cmake.in:
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) 2017-2020, Lawrence Livermore National Security, LLC and
```
Note: These new auto-generated host-configs for the examples capture information necessary to build the examples, like the compiler and paths to TPLs.

kennyweiss commented on src/thirdparty/CMakeLists.txt:
```diff
@@ -19,11 +19,11 @@ target_include_directories(cli11 SYSTEM INTERFACE
 install(DIRECTORY   ${PROJECT_SOURCE_DIR}/thirdparty/CLI11
         DESTINATION include )
-install(EXPORT cli11-targets DESTINATION lib/cmake)
+install(EXPORT axom-targets DESTINATION lib/cmake)
```
@white238 -- I changed this so that all exports are placed within a single ``axom-targets`` file.

rhornung67 commented on src/cmake/axom-config.cmake.in:
```diff
@@ -0,0 +1,69 @@
+
+@PACKAGE_INIT@
+
+if(NOT AXOM_FOUND)
```
🥇

kennyweiss commented on src/examples/using-with-cmake/CMakeLists.txt:
```diff
@@ -3,37 +3,77 @@
+if(AXOM_USE_CONDUIT)
+    check_var_and_path(CONDUIT_DIR ${CONDUIT_DIR}/lib/cmake/ConduitConfig.cmake)
+
+    find_package(Conduit REQUIRED
```
It looks like I *will* need to take care of this in ``axom-config``.  

I just tried a raja/umpire configuration and cmake complained about not knowing what ``-lraja`` and ``-lumpire`` is.

Also, I ran into the problem that @white238 mentioned yesterday for ``openmp+fortran`` configurations.  Without any workarounds, the user will have to enable ``Fortran`` or CMake complains about the generator expression:
```
CMake Error at CMakeLists.txt:77 (target_link_libraries):
  Error evaluating generator expression:

    $<COMPILE_LANGUAGE:Fortran>

  $<COMPILE_LANGUAGE:...> Unknown language.
```
See: https://github.com/LLNL/blt/issues/205

cyrush replied:
you do need to find package as part of the exported targets, might need the same for raja as well.

kennyweiss replied:
Done.

kennyweiss replied:
I updated the export/import process.  
The only necessary config variable is now `AXOM_DIR`
(and `BLT_SOURCE_DIR` if it is a blt-based build system).

gzagaris commented on src/cmake/axom-config.cmake.in:
```diff
@@ -0,0 +1,69 @@
+  # Set version and paths
+  #----------------------------------------------------------------------------
+  
+  set(AXOM_VERSION "@AXOM_VERSION_FULL@")
```
We should also publish the `MAJOR`, `MINOR` and `PATCH` variables since we have those. That way consumers of this can add conditional logic in their respective CMake and or BLT build system.

gzagaris commented on src/cmake/axom-config.cmake.in:
```diff
@@ -0,0 +1,69 @@
+  set(AXOM_VERSION "@AXOM_VERSION_FULL@")
+  
+  set(AXOM_INSTALL_PREFIX "@AXOM_INSTALL_PREFIX@")
+  set(AXOM_INCLUDE_DIRS "${AXOM_INSTALL_PREFIX}/include/axom")
```
Remove `axom` at the end, since in applications we prefix all of axom includes with `axom`. This would also make available the bundled TPLs that we install.

kennyweiss replied:
Thanks @gzagaris -- I agree, the `axom` part doesn't belong there.

gzagaris commented on src/cmake/axom-config.cmake.in:
```diff
@@ -0,0 +1,69 @@
+  set(AXOM_ENABLE_PRIMAL      "@AXOM_ENABLE_PRIMAL@")
+  set(AXOM_ENABLE_SPIN        "@AXOM_ENABLE_SPIN@")
+  set(AXOM_ENABLE_SIDRE       "@AXOM_ENABLE_SIDRE@")
+  set(AXOM_ENABLE_QUEST       "@AXOM_ENABLE_QUEST@")
```
I suggest we keep this list sorted alphabetically.

white238 replied:
But i like being first.... makes me feel important

gzagaris replied:
If you are referring to `LUMBERJACK` appearing first on the list, that would still be the case after the list is sorted! So, there you go!

white238 replied:
! i swear i know my alphabet! Though my next component will be AAA (Absolutely Awesome Acronym)

gzagaris commented on src/examples/using-with-cmake/example.cpp:
```diff
@@ -10,12 +10,20 @@
+   std::cout << fmt::format(
+        "Example of using and installed version of axom v{}.{}.{}-{}",
+        AXOM_VERSION_MAJOR, AXOM_VERSION_MINOR,
+        AXOM_VERSION_PATCH, AXOM_VERSION_EXTRA) << std::endl << std::endl;
```
You shouldn't have to use `fmt` for this. 

Given that the intent of this is to provide a simple example of how to link `Axom` to a downstream `CMake` application, I prefer we remove anything extraneous and keep things simple.

This could be simplified to:
```
std::cout << AXOM_VERSION_FULL << "-" << AXOM_VERSION_PATH << std::endl;
```

kennyweiss replied:
Thanks @gzagaris -- I intentionally added `fmt` to this example to ensure that it is being properly exported and as an example of how to use this.  (See PR description).

Side note: Once we address #209 the ``fmt`` usage will have to be guarded.

gzagaris replied:
OK -- maybe we should add advertisement banners in our examples too :)

That's fine, but technically it would be better if it was a separate test, e.g., `test-fmt-export.cpp` or something like that. I would imagine we'd want to have similar tests for all our other built-in TPLs as well, not just `fmt`.

cyrush commented on src/examples/using-with-cmake/CMakeLists.txt:
```diff
@@ -3,37 +3,77 @@
+        message (FATAL_ERROR "Path '${_variable}' does not contain required"
+                             " file '${_file}'.")
+    endif()
+endfunction()
```
this is handy, but a possible draw back is this looks like cmake vodoo and it's the first thing folks will have to process when trying to learn how to use Axom.

kennyweiss replied:
Thanks @cyrush -- I removed this function and streamlined these `CMakeLists.txt` files.

white238 commented on src/cmake/axom-config.cmake.in:
```diff
@@ -0,0 +1,69 @@
+  set(AXOM_ENABLE_SIDRE       "@AXOM_ENABLE_SIDRE@")
+  set(AXOM_ENABLE_QUEST       "@AXOM_ENABLE_QUEST@")
+
+  set(AXOM_USE_CLI11          "@AXOM_USE_CLI11@")
```
Maybe pull the built-in's into their own section as a nit-pick...

white238 commented on src/cmake/axom-config.cmake.in:
```diff
@@ -0,0 +1,69 @@
+  set(AXOM_SCR_DIR            "@SCR_DIR@")
+  set(AXOM_UMPIRE_DIR         "@UMPIRE_DIR@")
+
+  set(AXOM_DATA_DIR           "@AXOM_DATA_DIR@")
```
Do we install the data dir? Is that important to remember?

kennyweiss replied:
Not sure -- I thought users might want to point to that, if it was available.

But, since it is not installed as part of the axom installation, I think you're right. I'll remove it.

kennyweiss replied:
Done.

kennyweiss commented:
I updated the export/import process to bring in the required TPLs, so users now only need to supply `AXOM_DIR`.  Specifically, they no longer need to supply `CONDUIT_DIR`, `RAJA_DIR`, etc..., but we'll use their variables, if provided.

I also updated our llnl CI scripts to test the installed version of Axom with the `using-with-cmake` and `using-with-blt` examples.  

Due to complications with using a cuda-based configuration, I disabled the installed examples tests in these configurations for the time being.
We are also not currently testing the installed make-based example, which I think can be better handled if we export `pkg-config` files.
I'll create tickets for both of these.

kennyweiss commented on scripts/llnl_scripts/llnl_lc_build_tools.py:
```diff
@@ -386,6 +388,73 @@ def build_and_test_host_config(test_root,host_config):
     sexe("ls %s/lib" %     install_dir, echo=True, error_prefix="WARNING:")
     sexe("ls %s/bin" %     install_dir, echo=True, error_prefix="WARNING:")
+
+    # test the installation using installed cmake examples
```
Note: Our LLNL CI tests are now running the cmake and blt examples to verify a proper installation.

rhornung67 commented:
@kennyweiss this looks really good. Nice improvements! Thanks!

kennyweiss commented on src/cmake/axom-config.cmake.in:
```diff
@@ -0,0 +1,131 @@
+  get_filename_component(AXOM_CMAKE_CONFIG_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)
+  include(${AXOM_CMAKE_CONFIG_DIR}/axom-targets.cmake)
+
+  # HACK: Clear INTERFACE_COMPILE_OPTIONS to avoid requiring OpenMP in user code
```
@white238 -- This was a hack to remove openmp flags from our exported target.
This should hopefully be unnecessary after our handling of openmp flags is fixed/improved.

Bonus: Users no longer need to set `ENABLE_FORTRAN=ON` in their build systems just to get `C++` code to compile :)

gzagaris commented on src/cmake/axom-config.cmake.in:
```diff
@@ -0,0 +1,131 @@
+
+  # Language features
+  set(AXOM_ENABLE_FORTRAN     "@ENABLE_FORTRAN@")
+  set(AXOM_USE_CUDA           "@AXOM_USE_CUDA@")
```
We should probably publish `AXOM_CUDA_ARCH` here.  That way, downstream build systems can access that information and add sanity checks etc. This would help with #236

kennyweiss replied:
Let's handle that separately, once we settle on #236, and where this goes.

gzagaris replied:
I think irrespective of where the `CUDA_ARCH` is handled and where the value of `CUDA_ARCH` is coming from, we should have it in the `cmake-config`. At this point, we only have this information in `AXOM_CUDA_ARCH`. This can change in the future, but, I think it is useful to have this in the cmake-config, so that it can be made accessible in the build system of downstream libraries.

Please, create a Github issue regarding publishing `CUDA_ARCH` in our CMake config.

gzagaris replied:
See #242. Please, feel free to make additional comments / suggestions.

gzagaris commented on src/cmake/axom-config.cmake.in:
```diff
@@ -0,0 +1,131 @@
+    set(AXOM_CONDUIT_DIR  "@CONDUIT_DIR@")
+    if(NOT CONDUIT_DIR) 
+      set(CONDUIT_DIR ${AXOM_CONDUIT_DIR}) 
+    endif()
```
Is this if-logic here correct? Do we need both `AXOM_CONDUIT_DIR` and `CONDUIT_DIR`?

Couldn't we just do:
```
find_dependency(Conduit REQUIRED NO_DEFAULT_PATHS PATHS "${CONDUIT_DIR}/lib/cmake")
```

kennyweiss replied:
This logic was an attempt to prioritize a user-provided `FOO_DIR`, 
but if they don't provide it, fall back to axom's exported `AXOM_FOO_DIR`

I didn't want to require the user to specify any of these libraries if they don't want/need to.
For an example, please check out the `using-with-cmake` example, where the user only needs to specify `AXOM_DIR`, but not `CONDUIT_DIR`.

This tracks with what `conduit` and `ascent` were doing.

kennyweiss replied:
To clarify -- two things are going on here:
1. For each dependency `FOO`, we are importing `AXOM_FOO_DIR` into the user's build system (using a variable from axom's build system)
2. In some (but not all cases), we are also importing the CMake exports of package `FOO`, into the user's build system, using the `find_dependency()` macro.  The path that we're searching for that is either a user-provided `FOO_DIR` or the axom exported `AXOM_FOO_DIR`.

I think part of the confusion relates to where these variables are defined. 
Please note the distinction between the following:
* `@FOO_DIR@` -- this is evaluating axom's variable
* `FOO_DIR` -- this is a string for a variable in the user's build system
* `${FOO_DIR}` -- this is evaluating a variable in the user's build system.

For example, the `CONDUIT_DIR` variable in this file is from the user application's build system, while `@CONDUIT_DIR@` is the path to axom's conduit.

gzagaris replied:
I see -- But, `AXOM_<FOO>_DIR` and `<FOO>_DIR` should really be the same?

I am wondering if we should check that and throw an error?

kennyweiss replied:
I was thinking about that too.
Ideally, they'd be the same, and clearly, problems can arise when they're not, but I'm not sure that it is always a problem if they're different.

Let's leave it as is for now, and fix it down the road if/when problems arise.

cyrush replied:
Some folks move install dirs (to be clear *I don't* recommend this), but this could help mitigate some of the consequences when that is in play.

gzagaris replied:
I cannot really think of an instance where someone wants to use a one version of conduit with their application and point to Axom that uses a different version of conduit. I think allowing is bound to cause problems down the road.

Could you please create an issue so that we dont' forget.

cyrush replied:
my example:  it is the same install - but happens to have been relocated to a different place on the file system than that was encoded during a build. 

Folks are doing this now. Turning this into an error could result in many donut payments.
That said - for the know case of this I don't think they are including Axom via cmake

There are other tricky issues with relocating that can't magically be fixed. 
In ascent we have had to provide runtime overrides for folks who choose to move the install dir but still need to access assets (like html templates)

gzagaris replied:
> Folks are doing this now. Turning this into an error could result in many donut payments.
> That said - for the know case of this I don't think they are including Axom via cmake

I know exactly what you mean and I totally agree they *shouldn't be moving install directories* 

Also, in this case, Axom, Conduit, Ascent, etc. are not included via CMake.

But, there are instances, where I want to build my own version of Axom for development and point to those "installed" libraries that have been moved. We have been going in and modifying the paths manually in this case. I run into this fairly recently. Having additional guards would be helpful.

gzagaris replied:
I would also add that making donut payments for things that folks should not do in the first place encourages the wrong behavior.

So, let's give them an error instead and provide a one-time, generous, donut payment once they *stop* doing things they shouldn't be doing in the first place 😄

rhornung67 replied:
I recommend *not* advertising the donut reward for changing bad behavior into good. That may encourage certain people to do the wrong thing and then correct themselves just to get a donut.... 🙉

gzagaris replied:
@rhornung67 -- too late for that. That's something that has been going on for decades now...I am surprised it is not in the developer's manual (oh wait...you wrote that)

gzagaris replied:
See #243. Please, feel free to make additional  comments / suggestions there.

I'll get everyone donuts if we fix this 😄

gzagaris commented on src/cmake/axom-config.cmake.in:
```diff
@@ -0,0 +1,131 @@
+  endif()
+
+  # raja
+  if(AXOM_USE_RAJA)
```
SInce we are doing more or less the same thing for each TPL, we may want to consider having a macro, `axom_import_dependencies(...)`. Perhaps, it could be also made general enough to go in BLT.

kennyweiss replied:
Thanks @gzagaris.

At this point, I am not sure there is enough custom code to warrant a separate macro. These three happen to look similar, but other TPLs (e.g. mfem or hdf5, if we want/need to export it) would likely look different.

gzagaris replied:
That sounds fair.
